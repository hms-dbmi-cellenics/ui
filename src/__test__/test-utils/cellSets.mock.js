const mockCellSets = {
  initialLoadPending: false,
  error: false,
  loading: false,
  accessible: true,
  hierarchy: [
    {
      key: 'louvain',
      children: [{ key: 'cluster-a' }, { key: 'cluster-b' }, { key: 'cluster-c' }],
      cellIds: new Set(),
    },
    {
      key: 'sample',
      children: [{ key: 'sample-1' }, { key: 'sample-2' }],
      cellIds: new Set(),
    },
    {
      key: 'scratchpad',
      children: [{ key: 'scratchpad-a' }],
      cellIds: new Set(),
    },
  ],
  properties: {
    'cluster-a': {
      name: 'cluster a',
      key: 'cluster-a',
      cellIds: new Set([0, 1, 6]),
      color: '#01FFFF',
    },
    'cluster-b': {
      name: 'cluster b',
      key: 'cluster-b',
      cellIds: new Set([2, 3]),
      color: '#23FFFF',
    },
    'cluster-c': {
      name: 'cluster c',
      key: 'cluster-c',
      cellIds: new Set([4, 5]),
      color: '#45FFFF',
    },
    'sample-1': {
      name: 'Sample 1',
      key: 'sample-1',
      cellIds: new Set([0, 1, 2]),
      color: '#012FFF',
    },
    'sample-2': {
      name: 'Sample 2',
      key: 'sample-2',
      cellIds: new Set([3, 4, 5]),
      color: '#345FFF',
    },
    'scratchpad-a': {
      cellIds: new Set(['5']),
      key: 'scratchpad-a',
      name: 'New Cluster',
      color: '#5FFFFF',
    },
    louvain: {
      cellIds: new Set(),
      name: 'Louvain clusters',
      key: 'louvain',
      type: 'cellSets',
      rootNode: true,
    },
    sample: {
      cellIds: new Set([6, 7, 8]),
      name: 'Samples',
      key: 'sample',
      type: 'metadataCategorical',
      rootNode: true,
    },
    scratchpad: {
      cellIds: new Set(),
      name: 'Scratchpad',
      key: 'scratchpad',
      type: 'cellSets',
      rootNode: true,
    },
  },
};

export {
  // eslint-disable-next-line import/prefer-default-export
  mockCellSets,
};
