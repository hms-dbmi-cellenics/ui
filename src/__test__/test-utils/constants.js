export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
