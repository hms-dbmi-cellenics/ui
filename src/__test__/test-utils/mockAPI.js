import _ from 'lodash';

// Default platform data
import cellSetsData from '__test__/data/cell_sets.json';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';

import fake from '__test__/test-utils/constants';

import {
  responseData,
} from '__test__/test-utils/mockData';

const promiseResponse = (
  response,
  options = {},
) => Promise.resolve(new Response(response, options));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (
  mockedResults,
) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }

  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (experimentId) => ({
  [`experiments/${experimentId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === experimentId)),
  ),
  [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
    JSON.stringify(processingConfigData),
  ),
  [`experiments/${experimentId}/cellSets$`]: () => promiseResponse(
    JSON.stringify(cellSetsData),
  ),
  [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${experimentId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${experimentId}`]: () => statusResponse(200, 'OK'),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
};
