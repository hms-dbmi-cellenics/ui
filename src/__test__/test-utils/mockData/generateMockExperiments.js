import _ from 'lodash';
import fake from '__test__/test-utils/constants';

const experimentFromTemplate = (idx) => (
  {
    id: `${fake.EXPERIMENT_ID}-${idx}`,
    name: `${fake.EXPERIMENT_NAME}-${idx}`,
    description: `Mock experiment ${idx}`,
    samplesOrder: [],
    metadataKeys: [],
    pipelines: {
      qc: {
        shouldRerun: null,
        stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:stateMachine:biomage-qc-development-mock997586f1bae0a6d5c5beed65663ec151test',
        executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-qc-development-mock997586f1bae0a6d5c5beed65663ec1519de4:422493f3-68a0-4ecb-8b17-0e8cd7a3test',
      },
      gem2s: {
        shouldRerun: true,
        stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:stateMachine:biomage-gem2s-development-mock997586f1bae0a6d5c5beed65663ec151test',
        executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-gem2s-development-mock997586f1bae0a6d5c5beed65663ec1519de4:5f273383-f093-4112-95b0-d76db896test',
      },
      obj2s: {
        paramsHash: 'mock57f90e94eeaa82ee6fb7627110828f2etest',
        stateMachineArn: 'arn:aws:states:eu-west-1:000000000000:stateMachine:biomage-obj2s-development-mock997586f1bae0a6d5c5beed65663ec151test',
        executionArn: 'arn:aws:states:eu-west-1:000000000000:execution:biomage-obj2s-development-mock997586f1bae0a6d5c5beed65663ec1519de4:5f273383-f093-4112-95b0-d76db896test',
      },
    },
    notifyByEmail: true,
    pipelineVersion: 1,
    createdAt: fake.MOCK_DATETIME,
    updatedAt: fake.MOCK_DATETIME,
  }
);

const generateMockExperiments = (
  numExperiments = 1,
  attrs = [],
) => {
  const experimentsList = [];

  _.times(numExperiments, (idx) => {
    const experiment = _.merge(
      {},
      experimentFromTemplate(idx),
      attrs[idx] ?? {},
    );

    experimentsList.push(experiment);
  });

  return experimentsList;
};

export default generateMockExperiments;
