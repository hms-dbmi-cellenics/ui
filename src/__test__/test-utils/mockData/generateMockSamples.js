import _ from 'lodash';

import fake from '__test__/test-utils/constants';
import sampleFileType from 'utils/sampleFileType';

const mockSampleTemplate = (experimentId, sampleId, idx) => ({
  id: sampleId,
  experimentId,
  name: `Mock sample ${idx}`,
  sampleTechnology: '10x',
  createdAt: '2021-12-07 17:36:27.773+00',
  updatedAt: '2021-12-07 17:38:42.036+00',
  metadata: { age: 'BL', timePoint: 'BL' },
  files: {
    matrix10X: {
      uploadStatus: 'uploaded',
      sampleFileType: sampleFileType.MATRIX_10_X,
      size: 1000,
      s3Path: 'testcfd8122f-25af-4f1a-a306-3268d44ed401',
    },
    barcodes10X: {
      uploadStatus: 'uploaded',
      sampleFileType: sampleFileType.BARCODES_10_X,
      size: 100,
      s3Path: 'testcfd8122f-25af-4f1a-a306-3268d44ed402',
    },
    features10X: {
      uploadStatus: 'uploaded',
      sampleFileType: sampleFileType.FEATURES_10_X,
      size: 100,
      s3Path: 'testcfd8122f-25af-4f1a-a306-3268d44ed403',
    },
  },
  options: {},
});

const generateMockSamples = (experimentId, numSamples = 3) => {
  const samples = [];

  _.times(numSamples, (idx) => {
    const newSampleId = `${fake.SAMPLE_ID}-${idx}`;
    samples.push(mockSampleTemplate(experimentId, newSampleId, idx));
  });

  return samples;
};

export default generateMockSamples;
