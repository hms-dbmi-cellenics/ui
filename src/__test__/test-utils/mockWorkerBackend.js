import * as socketConnectionMocks from 'utils/socketConnection';
import * as fetchWorkResponseMocks from 'utils/work/fetchWork';

import SocketMock from 'socket.io-mock';

const socketMock = new SocketMock();

jest.mock('utils/work/fetchWork', () => {
  const mockSeekFromS3 = jest.fn();
  const originalModule = jest.requireActual('utils/work/fetchWork');

  return {
    __esModule: true, // Use it when dealing with esModules
    ...originalModule,
    fetchWork: mockSeekFromS3,
  };
});

jest.mock('aws-amplify', () => ({
  configure: () => ({
    Storage: {
      AWSS3: {
        bucket: 'biomage-originals-test',
      },
    },
  }),
  Storage: {
    get: async (ETag) => `http://mock.s3.amazonaws.com/${ETag}`,
  },
  Auth: {
    federatedSignIn: jest.fn(),
    signOut: jest.fn(),
    currentAuthenticatedUser: jest.fn().mockImplementation(async () => ({
      username: 'mockuser',
      attributes: {
        email: 'mock@user.name',
        name: 'Mocked User',
      },
    })),
    currentSession: () => ({
      getIdToken: (() => ({
        getJwtToken: () => 'fakeJwtToken',
      })),
    }),
  },
  Hub: {
    listen: jest.fn(),
  },
}));

jest.mock('utils/socketConnection', () => {
  const mockEmit = jest.fn();
  const mockOn = jest.fn();
  return {
    __esModule: true,
    default: new Promise((resolve) => {
      resolve({ emit: mockEmit, on: mockOn, id: '1234' });
    }),
    mockEmit,
    mockOn,
  };
});

// To mock worker response, modify the response returned from the API
// Set to null to force fetch from API
fetchWorkResponseMocks.fetchWork.mockImplementation(() => null);

// Set up socket emitter mock
socketConnectionMocks.mockEmit.mockImplementation((workRequestType, requestBody) => {
  const responseBody = { response: { error: false } };

  // After emitting, send reply to listener
  socketMock.socketClient.emit(`WorkResponse-${requestBody.ETag}`, responseBody);
});

// Set up socket listener mock
socketConnectionMocks.mockOn.mockImplementation((channel, f) => {
  socketMock.on(channel, (responseBody) => f(responseBody));
});
