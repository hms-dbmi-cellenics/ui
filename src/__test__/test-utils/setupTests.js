import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
  Spatial: () => 'Mocked Spatial',
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});
