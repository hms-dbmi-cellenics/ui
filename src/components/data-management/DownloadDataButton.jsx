import React, { useEffect, useState } from 'react';
import _ from 'lodash';
import {
  Tooltip, Dropdown, Button, Space,
} from 'antd';

import { useSelector, useDispatch } from 'react-redux';
import { saveAs } from 'file-saver';
import { ClipLoader } from 'react-spinners';

import downloadTypes from 'utils/data-management/downloadTypes';
import endUserMessages from 'utils/endUserMessages';
import pipelineStatus from 'utils/pipelineStatusValues';
import { exportQCParameters, filterQCParameters } from 'utils/data-management/exportQCParameters';

import { loadBackendStatus } from 'redux/actions/backendStatus/index';

import { getBackendStatus } from 'redux/selectors';
import handleError from 'utils/http/handleError';
import downloadProcessedMatrix from 'utils/extraActionCreators/downloadProcessedMatrix';

const DownloadDataButton = () => {
  const dispatch = useDispatch();
  const experimentSettings = useSelector((state) => state.experimentSettings);
  const experiments = useSelector((state) => state.experiments);
  const { activeExperimentId } = experiments.meta;
  const activeExperiment = experiments[activeExperimentId];

  const {
    status: backendStatuses, loading: backendLoading,
  } = useSelector(getBackendStatus(activeExperimentId));

  const samples = useSelector((state) => state.samples);
  const [pipelineHasRun, setPipelineHasRun] = useState(false);
  const [allSamplesAnalysed, setAllSamplesAnalysed] = useState(false);
  const [downloadingProcessedSeurat, setDownloadingProcessedSeurat] = useState(false);
  const [dropdownExpanded, setDropdownExpanded] = useState(false);

  useEffect(() => {
    if (activeExperimentId && !backendLoading && !backendStatuses) {
      dispatch(loadBackendStatus(activeExperimentId));
    }
  }, [activeExperimentId]);

  useEffect(() => {
    setPipelineHasRun(
      activeExperimentId && (backendStatuses?.pipeline?.status === pipelineStatus.SUCCEEDED
        || backendStatuses?.seurat?.status === pipelineStatus.SUCCEEDED),
    );
  }, [backendStatuses]);

  useEffect(() => {
    setAllSamplesAnalysed(getAllSamplesAnalysed());
  }, [activeExperiment, experimentSettings]);

  const getAllSamplesAnalysed = () => {
    // Returns true only if there is at least one sample in the currently active
    // project AND all samples in the project have been analysed.
    if (!activeExperiment?.sampleIds?.length) {
      return false;
    }
    const steps = Object.values(_.omit(experimentSettings?.processing, ['meta']));
    return steps.length > 0
      // eslint-disable-next-line no-prototype-builtins
      && activeExperiment?.sampleIds?.every((s) => steps[0].hasOwnProperty(s));
  };

  const downloadExperimentData = async (type) => {
    try {
      if (!activeExperimentId) throw new Error('No experimentId specified');
      if (!downloadTypes.has(type)) throw new Error('Invalid download type');

      setDownloadingProcessedSeurat(true);
      await dispatch(downloadProcessedMatrix(activeExperimentId));
      setDownloadingProcessedSeurat(false);
      setDropdownExpanded(false);
    } catch (e) {
      handleError(e, endUserMessages.ERROR_DOWNLOADING_DATA);
    }
  };

  const menuItems = [
    {
      key: 'download-processed-seurat',
      disabled: !pipelineHasRun || backendLoading,
      onClick: (e) => {
        e.domEvent.stopPropagation();
        downloadExperimentData('processed-matrix');
      },
      label: (
        <Tooltip
          title={
            pipelineHasRun
              ? 'With Data Processing filters and settings applied'
              : 'Launch analysis to process data'
          }
          placement='left'
        >
          <Space>
            Processed Seurat object (.rds)
            {downloadingProcessedSeurat && <ClipLoader size={20} color='#8f0b10' />}
          </Space>
        </Tooltip>
      ),
    },
    {
      key: 'download-processing-settings',
      disabled: !allSamplesAnalysed || backendLoading,
      onClick: () => {
        const config = _.omit(experimentSettings.processing, ['meta']);
        const filteredConfig = filterQCParameters(
          config, activeExperiment.sampleIds, samples,
        );
        const blob = exportQCParameters(filteredConfig);
        saveAs(blob, `${activeExperimentId.split('-')[0]}_settings.txt`);
      },
      label: (
        allSamplesAnalysed
          ? 'Data Processing settings (.txt)'
          : (
            <Tooltip title='One or more of your samples has yet to be analysed' placement='left'>
              Data Processing settings (.txt)
            </Tooltip>
          )
      ),
    },
  ];

  return (
    <Dropdown
      open={dropdownExpanded}
      onOpenChange={(visible) => setDropdownExpanded(visible)}
      trigger={['click']}
      menu={{
        items: menuItems,
        onClick: (e) => {
          if (e.key !== 'download-processed-seurat') setDropdownExpanded(false);
        },
      }}
      placement='bottomRight'
      disabled={
        experiments.ids.length === 0
        || activeExperiment.sampleIds.length === 0
      }
    >
      <Button>
        Download
      </Button>
    </Dropdown>

  );
};

export default React.memo(DownloadDataButton);
