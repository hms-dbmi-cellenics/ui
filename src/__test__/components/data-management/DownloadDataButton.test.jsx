import React from 'react';
import _ from 'lodash';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { saveAs } from 'file-saver';

import { screen, render, fireEvent } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import preloadAll from 'jest-next-dynamic';

import { act } from 'react-dom/test-utils';

import DownloadDataButton from 'components/data-management/DownloadDataButton';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import endUserMessages from 'utils/endUserMessages';

import { processingConfig } from '__test__/test-utils/mockData';
import backendStatusData from '__test__/data/backend_status.json';

import fetchWork from 'utils/work/fetchWork';

import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { makeStore } from 'redux/store';
import { loadExperiments } from 'redux/actions/experiments';

import mockAPI, { generateDefaultMockAPIResponses, promiseResponse } from '__test__/test-utils/mockAPI';
import fake from '__test__/test-utils/constants';
import { loadBackendStatus } from 'redux/actions/backendStatus';
import { loadProcessingSettings } from 'redux/actions/experimentSettings';
import downloadFromUrl from 'utils/downloadFromUrl';
import writeToFileURL from 'utils/upload/writeToFileURL';

jest.mock('file-saver');

jest.mock('utils/work/fetchWork');
jest.mock('utils/pushNotificationMessage');
jest.mock('utils/downloadFromUrl');
jest.mock('utils/upload/writeToFileURL');

const experimentId = `${fake.EXPERIMENT_ID}-0`;

enableFetchMocks();

const mockAPIResponse = _.merge(
  generateDefaultMockAPIResponses(experimentId),
  {
    [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
      JSON.stringify({ processingConfig }),
    ),
  },
);

describe('DownloadDataButton', () => {
  let storeState;

  beforeAll(async () => {
    await preloadAll();
  });

  beforeEach(() => {
    jest.clearAllMocks(); // Do not mistake with resetAllMocks()!
    fetchMock.resetMocks();

    storeState = makeStore();
  });

  const renderDownloadDataButton = async () => {
    await act(async () => {
      storeState.dispatch(loadExperiments());
    });
    await act(async () => {
      storeState.dispatch(loadProcessingSettings(experimentId));
    });
    await act(async () => {
      storeState.dispatch(loadBackendStatus(experimentId));
    });

    await act(async () => {
      render(
        <Provider store={storeState}>
          <DownloadDataButton />
        </Provider>,
      );
    });
  };

  const getMenuItems = async () => {
    const menu = await screen.getByText('Download');
    expect(menu).not.toBeDisabled();

    await act(async () => {
      userEvent.click(menu);
    });

    const options = await screen.getAllByRole('menuitem');
    return options;
  };

  it('should render the download data menu', async () => {
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    await renderDownloadDataButton();

    const options = await getMenuItems();
    expect(options).toHaveLength(2);
    expect(options[0]).toHaveTextContent('Processed Seurat object (.rds)');
    expect(options[1]).toHaveTextContent('Data Processing settings (.txt)');
    expect(options[0]).toHaveAttribute('aria-disabled', 'false');
    expect(options[1]).toHaveAttribute('aria-disabled', 'false');
  });

  it('Processed Seurat object option is disabled if qc has not succeeded', async () => {
    const backendStatus = _.cloneDeep(backendStatusData);

    backendStatus.pipeline.status = 'FAILED';
    backendStatus.gem2s.status = 'SUCCEEDED';

    const qcFailMockAPIResponse = _.merge(
      {},
      mockAPIResponse,
      { [`experiments/${experimentId}/backendStatus$`]: () => promiseResponse(JSON.stringify(backendStatus)) },
    );

    fetchMock.mockIf(/.*/, mockAPI(qcFailMockAPIResponse));

    await renderDownloadDataButton();
    const options = await getMenuItems();
    expect(options[0]).toHaveAttribute('aria-disabled', 'true');
    expect(options[1]).toHaveAttribute('aria-disabled', 'false');
  });

  it('Data procesing settings option is disabled if a step misses a sample', async () => {
    const processingConfigMissingSample = _.cloneDeep(processingConfig);
    // Remove settings for one sample
    delete processingConfigMissingSample.cellSizeDistribution[`${fake.SAMPLE_ID}-0`];

    const stepMissingMockAPIResponse = _.merge(
      {},
      mockAPIResponse,
      {
        [`experiments/${experimentId}/processingConfig$`]: () => promiseResponse(
          JSON.stringify({ processingConfig: processingConfigMissingSample }),
        ),
      },
    );

    fetchMock.mockIf(/.*/, mockAPI(stepMissingMockAPIResponse));

    await renderDownloadDataButton();

    const options = await getMenuItems();
    expect(options[0]).toHaveAttribute('aria-disabled', 'false');
    expect(options[1]).toHaveAttribute('aria-disabled', 'true');
  });

  it('Downloads processing config properly', async () => {
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    await renderDownloadDataButton();

    // Open the Download dropdown
    userEvent.click(screen.getByText(/Download/i));

    const downloadButton = screen.getByText(/Data Processing settings/i);

    await act(async () => {
      fireEvent.click(downloadButton);
    });

    expect(saveAs).toHaveBeenCalled();
  });

  it('Downloads processed matrix properly', async () => {
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    const mockResult = '--ImAMock--';
    const mockFileUrl = '--fileUrl--';

    fetchWork.mockImplementation((expId, body) => {
      if (body.name === 'GetEmbedding') {
        return Promise.resolve();
      }
      if (body.name === 'DownloadAnnotSeuratObject') {
        return Promise.resolve(mockResult);
      }
    });

    writeToFileURL.mockImplementationOnce(() => mockFileUrl);

    await renderDownloadDataButton();

    // Open the Download dropdown
    userEvent.click(screen.getByText(/Download/i));

    const downloadButton = screen.getByText(/Processed Seurat object/i);

    await act(async () => {
      fireEvent.click(downloadButton);
    });

    expect(fetchWork).toHaveBeenCalledTimes(2);
    expect(fetchWork.mock.calls).toMatchSnapshot();

    expect(writeToFileURL).toHaveBeenCalledWith(mockResult);
    expect(downloadFromUrl).toHaveBeenCalledWith(mockFileUrl, `${experimentId}_processed_matrix.rds`);
  });

  it('Shows an error if there is an error downloading data', async () => {
    fetchMock.mockIf(/.*/, mockAPI(mockAPIResponse));

    fetchWork.mockImplementation((expId, body) => {
      if (body.name === 'GetEmbedding') {
        return Promise.resolve();
      }

      if (body.name === 'DownloadAnnotSeuratObject') {
        return Promise.reject(new Error('some worker error'));
      }
    });

    await renderDownloadDataButton();

    // Open the Download dropdown
    userEvent.click(screen.getByText(/Download/i));

    const downloadButton = screen.getByText(/Processed Seurat object/i);

    await act(async () => {
      fireEvent.click(downloadButton);
    });

    expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
    expect(pushNotificationMessage).toHaveBeenCalledWith('error', endUserMessages.ERROR_DOWNLOADING_SEURAT_OBJECT);
  });

  it('Has options disabled if backend status is still loading', async () => {
    const loadingStatusApiResponse = _.merge(
      {},
      mockAPIResponse,
      { [`experiments/${experimentId}/backendStatus$`]: () => new Promise(() => { }) },
    );

    fetchMock.mockIf(/.*/, mockAPI(loadingStatusApiResponse));

    await renderDownloadDataButton();
    const options = await getMenuItems();

    expect(options[0]).toHaveAttribute('aria-disabled', 'true');
    expect(options[1]).toHaveAttribute('aria-disabled', 'true');
  });
});
