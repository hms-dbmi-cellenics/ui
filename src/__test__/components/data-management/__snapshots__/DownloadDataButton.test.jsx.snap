// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadDataButton Downloads processed matrix properly 1`] = `
[
  [
    "testae48e318dab9a1bd0bexperiment-0",
    {
      "config": {
        "distanceMetric": "cosine",
        "minimumDistance": 0.3,
      },
      "name": "GetEmbedding",
      "type": "umap",
      "useSaved": undefined,
    },
    [Function],
    [Function],
    {
      "timeout": 900,
    },
  ],
  [
    "testae48e318dab9a1bd0bexperiment-0",
    {
      "embeddingMethod": "umap",
      "name": "DownloadAnnotSeuratObject",
    },
    [Function],
    [Function],
    {
      "timeout": 180,
    },
  ],
]
`;
